idea/src/org/jetbrains/kotlin/idea/codeInsight/surroundWith/statement/KotlinStatementsSurrounder.java

```diff
@@ -24,12 +24,36 @@
 import com.intellij.util.IncorrectOperationException;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import org.jetbrains.kotlin.idea.caches.resolve.ResolutionUtils;
+import org.jetbrains.kotlin.psi.KtElement;
+import org.jetbrains.kotlin.psi.KtExpression;
+import org.jetbrains.kotlin.resolve.BindingContext;
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode;
 
 public abstract class KotlinStatementsSurrounder implements Surrounder {
 
     @Override
     public boolean isApplicable(@NotNull PsiElement[] elements) {
-        return elements.length > 0;
+        if (elements.length == 0) {
+            return false;
+        }
+
+        if (elements.length == 1 || !(elements[0] instanceof KtExpression) && !isApplicableWhenUsedAsExpression()) {
+            KtElement expression = (KtElement) elements[0];
```

Comment:
I now have a base class `KotlinControlFlowExpressionSurrounderBase` for both `try` and `if` expressions. I think the cost of `analyze` shouldn't be too bad because it is only called when the user invokes the "Sourround with" action, not on every key stroke.

Reply:
I have found that there was an analyze call for getting expression type in `KotlinExpressionSurrounder.isApplicable` sing long ago. Ok than :)